# Data Manipulation
## INSTRUCTIONS 

Every learner should submit his/her own homework solutions. However, you are allowed to discuss the homework with each other– but everyone must submit his/her own solution; you may not copy someone else’s solution. 

Follow the prompts in the attached jupyter notebook. Download the data and place it in your working directory, or modify the path to upload it to your notebook. Add markdown cells to your analysis to include your solutions, comments, answers. Add as many cells as you need, for easy readability comment when possible.Make sure that you run your whole notebook before saving and sending it to us.

Hopefully this homework will help you develop skills, make you understand data manipulation techniques.

Submission: Send in both a ipynb and a html/pdf file of your work. 

Good luck!


# Cleaning, wrangling data

**Data cleaning focuses on removing inaccurate data from your data set whereas data wrangling focuses on transforming the data's format, typically by converting “raw” data into another format more suitable for use.
This excersize uses the traffic_cameras file. Your task is to follow prompts to change, modify your data. Try your best!**

In [174]:
#read in libraries
import numpy as np
from sklearn.datasets import load_iris
from sklearn import preprocessing
import pandas as pd
 

In [175]:
# Reading the CSV file
df = pd.read_csv("traffic_cameras.csv")
 
# Printing top 5 rows
df.head()

Camera ID                              Location Name Camera Status  \
0        370   PLEASANT VALLEY RD / NUCKOLS CROSSING RD     TURNED_ON   
1        379             BARTON SPRINGS RD / KINNEY AVE     TURNED_ON   
2        404             SPRINGDALE RD / OAK SPRINGS DR     TURNED_ON   
3        447                 BRAKER LN / STONELAKE BLVD     TURNED_ON   
4        552              EXPOSITION BLVD / WESTOVER RD     TURNED_ON   

  Turn on Date Camera Manufacturer ATD Location ID Landmark  \
0    5/24/2018             Advidia    LOC16-003180      NaN   
1    5/21/2018             Advidia    LOC16-000640      NaN   
2     6/7/2018             Advidia    LOC16-000800      NaN   
3     9/9/2016             Advidia    LOC16-003740      NaN   
4    2/24/2020             Advidia    LOC16-003710      NaN   

  Signal Engineer Area Council District         Jurisdiction  ...  \
0            SOUTHEAST                2  AUSTIN FULL PURPOSE  ...   
1            SOUTHWEST                5  AUSTIN FULL PURPOSE  ...   
2            NORTHEAST             3, 1  AUSTIN FULL PURPOSE  ...   
3            NORTHWEST                7  AUSTIN FULL PURPOSE  ...   
4              CENTRAL               10  AUSTIN FULL PURPOSE  ...   

  CROSS_ST_AKA  COA Intersection ID                 Modified Date  \
0          NaN            5157009.0  10/28/2021 08:40:00 AM +0000   
1          NaN            5154418.0  10/29/2021 08:45:00 AM +0000   
2          NaN            5163165.0  10/29/2021 07:38:00 PM +0000   
3          NaN            5156985.0  10/29/2021 07:49:00 PM +0000   
4          NaN            5153874.0  10/29/2021 07:47:00 PM +0000   

   IP Comm Status  IP Comm Status Date and Time  Published Screenshots  \
0          ONLINE  10/28/2021 08:30:00 AM +0000                    NaN   
1          ONLINE  10/29/2021 08:35:00 AM +0000                    NaN   
2          ONLINE  10/28/2021 08:35:00 AM +0000                    NaN   
3          ONLINE  10/23/2021 08:35:00 AM +0000                    NaN   
4          ONLINE  10/20/2021 08:35:00 AM +0000                    NaN   

                             Screenshot Address          Funding  \
0  https://cctv.austinmobility.io/image/370.jpg         1/4 Cent   
1  https://cctv.austinmobility.io/image/379.jpg         1/4 Cent   
2  https://cctv.austinmobility.io/image/404.jpg         1/4 Cent   
3  https://cctv.austinmobility.io/image/447.jpg         1/4 Cent   
4  https://cctv.austinmobility.io/image/552.jpg  None Identified   

                         ID                        Location  
0  591a10a020eacf2d16669b94  POINT (-97.7449036 30.1844883)  
1  591a10a020eacf2d16669ba6    POINT (-97.761467 30.261982)  
2  591a10a120eacf2d16669bd8  POINT (-97.6904221 30.2735615)  
3  591a10a320eacf2d16669c2e  POINT (-97.7392883 30.3989582)  
4  5aa6bb0121cbcf4b8b767294  POINT (-97.7643051 30.3030338)  

[5 rows x 28 columns]

### 1. How many rows and columns does your data have?

In [176]:
### Your code goes here ###
df.shape

(802, 28)

##Your Answer goes here ### change the cell to markdown
row #: 802
column #: 28 

### 2. What can you tell us about the type of variables we have?

In [177]:
### Your code goes here ###
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 802 entries, 0 to 801
Data columns (total 28 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Camera ID                     802 non-null    int64  
 1   Location Name                 802 non-null    object 
 2   Camera Status                 802 non-null    object 
 3   Turn on Date                  442 non-null    object 
 4   Camera Manufacturer           646 non-null    object 
 5   ATD Location ID               802 non-null    object 
 6   Landmark                      94 non-null     object 
 7   Signal Engineer Area          799 non-null    object 
 8   Council District              790 non-null    object 
 9   Jurisdiction                  799 non-null    object 
 10  Location Type                 802 non-null    object 
 11  Primary St Segment ID         0 non-null      float64
 12  Cross St Segment ID           0 non-null      float64
 13  Prima

### Your asnwer should go in here ### change the cell to markdown
data types: float64(8), int64(1), object(19)



### 3. Delete only the columns that have all null values, name it df1 (nothing else, but null)

In [178]:
### Your code goes here ###
df1 = df.dropna(axis=1, how='all')

In [179]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 802 entries, 0 to 801
Data columns (total 23 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Camera ID                     802 non-null    int64  
 1   Location Name                 802 non-null    object 
 2   Camera Status                 802 non-null    object 
 3   Turn on Date                  442 non-null    object 
 4   Camera Manufacturer           646 non-null    object 
 5   ATD Location ID               802 non-null    object 
 6   Landmark                      94 non-null     object 
 7   Signal Engineer Area          799 non-null    object 
 8   Council District              790 non-null    object 
 9   Jurisdiction                  799 non-null    object 
 10  Location Type                 802 non-null    object 
 11  Primary Street Block          800 non-null    float64
 12  Primary Street                801 non-null    object 
 13  Cross

### 4. Dropp columns that have (any) null values name it df2

In [180]:
### Your code goes here ###
df2 = df.dropna(axis=1)

In [181]:
df2.info()
df2.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 802 entries, 0 to 801
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   Camera ID                     802 non-null    int64 
 1   Location Name                 802 non-null    object
 2   Camera Status                 802 non-null    object
 3   ATD Location ID               802 non-null    object
 4   Location Type                 802 non-null    object
 5   Modified Date                 802 non-null    object
 6   IP Comm Status                802 non-null    object
 7   IP Comm Status Date and Time  802 non-null    object
 8   Screenshot Address            802 non-null    object
 9   ID                            802 non-null    object
 10  Location                      802 non-null    object
dtypes: int64(1), object(10)
memory usage: 69.1+ KB


(802, 11)

### 5. Rename column names in df2 so they are more usable (name the new dataframe df3) to the followings: cam_id, loc_name, cam_stat, atd_loc_id, loc_type, date, comm_stat, comm_stat_date, screen_addr, id, location

In [182]:
### Your code goes here ###
df3 = df2.rename(columns={'Camera ID': 'cam_id', 
                          'Location Name': 'loc_name', 
                          'Camera Status': 'cam_stat', 
                          'ATD Location ID': 'atd_loc_id',
                          'Location Type': 'loc_type',
                          'Modified Date': 'date',
                          'IP Comm Status': 'comm_stat',
                          'IP Comm Status Date and Time': 'comm_stat_date',
                          'Screenshot Address': 'screen_addr',
                          'ID': 'id',
                          'Location': 'location',
                         })

In [183]:
df3.head

<bound method NDFrame.head of      cam_id                                           loc_name   cam_stat  \
0       370           PLEASANT VALLEY RD / NUCKOLS CROSSING RD  TURNED_ON   
1       379                     BARTON SPRINGS RD / KINNEY AVE  TURNED_ON   
2       404                     SPRINGDALE RD / OAK SPRINGS DR  TURNED_ON   
3       447                         BRAKER LN / STONELAKE BLVD  TURNED_ON   
4       552                      EXPOSITION BLVD / WESTOVER RD  TURNED_ON   
..      ...                                                ...        ...   
797    1190                             GUADALUPE ST / 46TH ST  TURNED_ON   
798    1274                    CESAR CHAVEZ ST / SAN MARCOS ST  TURNED_ON   
799    1275                     BURNET RD / RESEARCH BLVD SVRD  TURNED_ON   
800    1276   BASTROP HWY / MONTOPOLIS TO BASTROP NB RAMP (...  TURNED_ON   
801    1277                       BURNET RD / BRIGHT VERDE WAY  TURNED_ON   

       atd_loc_id loc_type                   

### 6. Split "date" column into two  new columns within df3 ('Dates' and 'Time') /modify df3 data/

In [184]:
### Your code goes here ###
split_data = df3['date'].str.split(' ', expand=True)

df3['Dates'] = split_data[0]
df3['Time'] = split_data[1] + ' ' + split_data[2] + ' ' + split_data[3]

del df3['date']

In [185]:
df3.head

<bound method NDFrame.head of      cam_id                                           loc_name   cam_stat  \
0       370           PLEASANT VALLEY RD / NUCKOLS CROSSING RD  TURNED_ON   
1       379                     BARTON SPRINGS RD / KINNEY AVE  TURNED_ON   
2       404                     SPRINGDALE RD / OAK SPRINGS DR  TURNED_ON   
3       447                         BRAKER LN / STONELAKE BLVD  TURNED_ON   
4       552                      EXPOSITION BLVD / WESTOVER RD  TURNED_ON   
..      ...                                                ...        ...   
797    1190                             GUADALUPE ST / 46TH ST  TURNED_ON   
798    1274                    CESAR CHAVEZ ST / SAN MARCOS ST  TURNED_ON   
799    1275                     BURNET RD / RESEARCH BLVD SVRD  TURNED_ON   
800    1276   BASTROP HWY / MONTOPOLIS TO BASTROP NB RAMP (...  TURNED_ON   
801    1277                       BURNET RD / BRIGHT VERDE WAY  TURNED_ON   

       atd_loc_id loc_type comm_stat         

### 7. Split atd_loc into two new columns 'Loc' and 'code' within df3

In [186]:
### Your code goes here ###
df3[['Loc', 'code']] = df3['atd_loc_id'].str.split('-', expand=True)
del df3['atd_loc_id']

In [187]:
df3.head

<bound method NDFrame.head of      cam_id                                           loc_name   cam_stat  \
0       370           PLEASANT VALLEY RD / NUCKOLS CROSSING RD  TURNED_ON   
1       379                     BARTON SPRINGS RD / KINNEY AVE  TURNED_ON   
2       404                     SPRINGDALE RD / OAK SPRINGS DR  TURNED_ON   
3       447                         BRAKER LN / STONELAKE BLVD  TURNED_ON   
4       552                      EXPOSITION BLVD / WESTOVER RD  TURNED_ON   
..      ...                                                ...        ...   
797    1190                             GUADALUPE ST / 46TH ST  TURNED_ON   
798    1274                    CESAR CHAVEZ ST / SAN MARCOS ST  TURNED_ON   
799    1275                     BURNET RD / RESEARCH BLVD SVRD  TURNED_ON   
800    1276   BASTROP HWY / MONTOPOLIS TO BASTROP NB RAMP (...  TURNED_ON   
801    1277                       BURNET RD / BRIGHT VERDE WAY  TURNED_ON   

    loc_type comm_stat                comm_st

### 8. What are the unique values in loc_type?

In [188]:
### Your code goes here ###
df3.loc_type.unique()

array(['ROADWAY', 'BUILDING'], dtype=object)

### Your asnwer should go in here ### change the cell to markdown
'ROADWAY' and 'BUILDING'

### 9. Replace 'ROADWAY' to '0',  'BUILDING' to '1' in the loc_type column within df3

In [189]:
### Your code goes here ###
df3['loc_type'] = df3['loc_type'].replace({'ROADWAY': '0', 'BUILDING': '1'})

In [190]:
df3.head

<bound method NDFrame.head of      cam_id                                           loc_name   cam_stat  \
0       370           PLEASANT VALLEY RD / NUCKOLS CROSSING RD  TURNED_ON   
1       379                     BARTON SPRINGS RD / KINNEY AVE  TURNED_ON   
2       404                     SPRINGDALE RD / OAK SPRINGS DR  TURNED_ON   
3       447                         BRAKER LN / STONELAKE BLVD  TURNED_ON   
4       552                      EXPOSITION BLVD / WESTOVER RD  TURNED_ON   
..      ...                                                ...        ...   
797    1190                             GUADALUPE ST / 46TH ST  TURNED_ON   
798    1274                    CESAR CHAVEZ ST / SAN MARCOS ST  TURNED_ON   
799    1275                     BURNET RD / RESEARCH BLVD SVRD  TURNED_ON   
800    1276   BASTROP HWY / MONTOPOLIS TO BASTROP NB RAMP (...  TURNED_ON   
801    1277                       BURNET RD / BRIGHT VERDE WAY  TURNED_ON   

    loc_type comm_stat                comm_st

In [191]:
df3.loc_type.unique()

array(['0', '1'], dtype=object)

### 10. Split the loc_name column into two new variables 'corner1', 'corner2' on '/'

In [192]:
### Your code goes here ###
df3[['corner1', 'corner2']] = df3['loc_name'].str.split('/', n=1, expand=True)
del df3['loc_name']

In [193]:
df3.head

<bound method NDFrame.head of      cam_id   cam_stat loc_type comm_stat                comm_stat_date  \
0       370  TURNED_ON        0    ONLINE  10/28/2021 08:30:00 AM +0000   
1       379  TURNED_ON        0    ONLINE  10/29/2021 08:35:00 AM +0000   
2       404  TURNED_ON        0    ONLINE  10/28/2021 08:35:00 AM +0000   
3       447  TURNED_ON        0    ONLINE  10/23/2021 08:35:00 AM +0000   
4       552  TURNED_ON        0    ONLINE  10/20/2021 08:35:00 AM +0000   
..      ...        ...      ...       ...                           ...   
797    1190  TURNED_ON        0    ONLINE  06/08/2021 08:30:00 AM +0000   
798    1274  TURNED_ON        0   OFFLINE  05/12/2021 08:30:00 AM +0000   
799    1275  TURNED_ON        0    ONLINE  03/13/2021 09:35:00 AM +0000   
800    1276  TURNED_ON        0    ONLINE  10/31/2021 08:30:00 AM +0000   
801    1277  TURNED_ON        0    ONLINE  06/17/2021 08:30:00 AM +0000   

                                       screen_addr                   